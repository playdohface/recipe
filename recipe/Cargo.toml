[package]
name = "recipe"
version = "0.1.0"
edition = "2021"

[dependencies]
convert_case = "0.6.0"
nom = "7.1.3"
nom_locate = "4.2.0"
serde_json = "1.0.132"
json5 = "0.4.1"
anyhow = "1.0.93"
