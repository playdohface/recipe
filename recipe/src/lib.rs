
pub mod context;
pub mod loader;
pub mod parser;
