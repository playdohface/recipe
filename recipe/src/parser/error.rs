use super::{Span, Token};

#[derive(Debug)]
pub enum ParseError<'a> {
    Generic(Span<'a>),
    InvalidToken(Token<'a>),
    InvalidCodeBlock(Span<'a>),
    NomError(nom::error::VerboseError<Span<'a>>),
}
