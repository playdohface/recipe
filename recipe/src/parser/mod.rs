use std::ops::RangeFrom;

use nom::{
    branch::alt,
    bytes::complete::{is_not, tag, take, take_till, take_until, take_while, take_while1},
    character::{
        complete::{char, line_ending, space0, space1},
        is_newline,
    },
    combinator::{fail, verify},
    error::{Error, ParseError},
    multi::many1_count,
    sequence::{delimited, preceded, terminated},
    IResult, InputIter, InputLength, InputTake, Parser, Slice,
};
use nom_locate::{position, LocatedSpan};

type Span<'a> = LocatedSpan<&'a str>;

#[derive(Debug, PartialEq, Eq)]
pub struct Heading<'a> {
    level: usize,
    text: &'a str,
}

#[derive(Debug, PartialEq, Eq)]
pub struct CodeBlock<'a> {
    executor: Option<&'a str>,
    name: Option<&'a str>,
    type_hint: Option<&'a str>,
    annotations: Vec<&'a str>,
    code: &'a str,
}

#[derive(Debug, PartialEq, Eq)]
pub struct Link<'a> {
    text: &'a str,
    path: &'a str,
}

#[derive(Debug, PartialEq, Eq)]
pub struct ToDirective<'a> {
    callpath: Vec<&'a str>,
    args: Vec<&'a str>,
}

#[derive(Debug, PartialEq, Eq)]
pub enum TokenType<'a> {
    Heading(Heading<'a>),
    Ident(&'a str),
    CodeBlock(CodeBlock<'a>),
    Link(Link<'a>),
    ToDirective(ToDirective<'a>),
}

#[derive(Debug, PartialEq, Eq)]
pub struct Token<'a> {
    position: Span<'a>,
    inner: TokenType<'a>,
}

impl<'a> Token<'a> {
    pub fn inner(&self) -> &TokenType<'a> {
        &self.inner
    }
}

#[derive(Debug)]
pub struct RecipeParser<'a> {
    src: Span<'a>,
}

impl<'a> RecipeParser<'a> {
    pub fn from_str(src: &'a str) -> Self {
        RecipeParser { src: src.into() }
    }
}
impl<'a> Iterator for RecipeParser<'a> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        if let Ok((rest, found)) = find_next(any_token)(self.src) {
            self.src = rest;
            Some(found)
        } else {
            None
        }
    }
}

fn any_token(inp: Span) -> IResult<Span, Token> {
    alt((
        parse_heading,
        parse_link,
        parse_code_block,
        parse_to_directive,
        parse_single_ident,
    ))(inp)
}

fn is_allowed_in_ident(c: char) -> bool {
    !c.is_whitespace() && c != '`'
}

fn is_allowed_in_name(c: char) -> bool {
    is_allowed_in_ident(c) && !['(', ')', '.', '[', ']', '{', '}'].contains(&c)
}

fn is_allowed_in_link_text(c: char) -> bool {
    !is_newline(c.try_into().unwrap_or_default()) && c != ']'
}

fn is_allowed_in_link_path(c: char) -> bool {
    !is_newline(c.try_into().unwrap_or_default()) && c != ')'
}

/// To-Directive: Starts with keyword "To" followed by at least one space.
/// finds valid names in backticks (\`name\`) until end of line (list of names is the callpath).
/// Optionally specifies required arguments as valid names preceded by 2 dashes(\`--name\`).
/// Callpath must be complete before the first argument, and must have at least one name in callpath before newline.
fn parse_to_directive(inp: Span) -> IResult<Span, Token, Error<LocatedSpan<&str>>> {
    let (inp, position) = position(inp)?;
    let (inp, _) = terminated(tag("To"), space1)(inp)?;
    let (mut inp, (callpath, end)) = find_many_till(
        delimited(tag("`"), valid_name, tag("`")),
        alt((tag("`--"), line_ending)),
        inp,
    )?;
    if callpath.is_empty() {
        return fail(position); // no elements in callpath means invalid directive
    }
    let mut args = Vec::new();
    if *end.fragment() == "`--" {
        let (rest, arg1) = terminated(valid_name, tag("`"))(inp)?;
        let (rest, (otherargs, _)) = find_many_till(
            delimited(tag("`--"), valid_name, tag("`")),
            line_ending,
            rest,
        )?;
        args.push(arg1);
        args.extend(otherargs);
        inp = rest;
    }

    let to_directive = ToDirective {
        callpath: callpath.into_iter().map(|e| *e.fragment()).collect(),
        args: args.into_iter().map(|e| *e.fragment()).collect(),
    };

    Ok((
        inp,
        Token {
            position,
            inner: TokenType::ToDirective(to_directive),
        },
    ))
}

fn parse_link(inp: Span) -> IResult<Span, Token> {
    let (inp, position) = position(inp)?;
    let (inp, text) = delimited(tag("["), take_while(is_allowed_in_link_text), tag("]"))(inp)?;
    let (inp, path) = delimited(tag("("), take_while(is_allowed_in_link_path), tag(")"))(inp)?;
    let link = TokenType::Link(Link {
        text: &text,
        path: &path,
    });
    Ok((
        inp,
        Token {
            position,
            inner: link,
        },
    ))
}

fn parse_code_block(inp: Span) -> IResult<Span, Token> {
    let (inp, position) = position(inp)?;
    let (inp, _) = tag("```")(inp)?;
    let (inp, executor) = valid_name(inp)?;
    let (inp, name) = preceded(space1, valid_name)(inp)?;
    let (mut inp, type_hint) = preceded(space0, delimited(tag("("), valid_name, tag(")")))(inp)?;
    let mut annotations: Vec<&str> = Vec::new();
    while let Ok((inp_ok, annotation)) = preceded(space1, valid_name)(inp) {
        annotations.push(&annotation);
        inp = inp_ok;
    }
    let (inp, _) = preceded(space0, line_ending)(inp)?;
    let (inp, code) = take_until("```")(inp)?;
    let (inp, _) = tag("```")(inp)?;
    let executor: Option<&str> = if executor.len() > 0 {
        Some(&executor)
    } else {
        None
    };
    let name: Option<&str> = if name.len() > 0 { Some(&name) } else { None };
    let type_hint: Option<&str> = if type_hint.len() > 0 {
        Some(&type_hint)
    } else {
        None
    };
    let codeblock = TokenType::CodeBlock(CodeBlock {
        executor,
        name,
        annotations,
        type_hint,
        code: &code,
    });
    Ok((
        inp,
        Token {
            position,
            inner: codeblock,
        },
    ))
}

fn parse_single_ident(inp: Span) -> IResult<Span, Token> {
    let (inp, position) = position(inp)?;
    let (inp, name) = delimited(tag("`"), take_while(is_allowed_in_ident), tag("`"))(inp)?;
    Ok((
        inp,
        Token {
            position,
            inner: TokenType::Ident(&name),
        },
    ))
}

fn parse_heading(inp: Span) -> IResult<Span, Token> {
    let (inp, position) = position(inp)?;
    let (inp, level) = many1_count(char('#'))(inp)?;
    let (inp, _) = take_till(|c: char| !c.is_whitespace())(inp)?;
    let (inp, text) = is_not("\r\n")(inp)?;
    let (inp, _) = line_ending(inp)?;

    let heading = Heading {
        level,
        text: text.trim_end(),
    };
    Ok((
        inp,
        Token {
            position,
            inner: TokenType::Heading(heading),
        },
    ))
}

fn is_allowed_at_name_start(c: char) -> bool {
    is_allowed_in_name(c) && !['-'].contains(&c)
}

fn is_valid_name<'a>(name: &Span) -> bool {
    name.starts_with(is_allowed_at_name_start)
}

fn valid_name<'a>(inp: Span<'a>) -> IResult<Span<'a>, Span<'a>> {
    verify(take_while1(is_allowed_in_name), is_valid_name)(inp)
}

fn find_next<I, O, G>(mut g: G) -> impl FnMut(I) -> IResult<I, O>
where
    G: FnMut(I) -> IResult<I, O>,
    I: Clone + InputTake + InputLength + InputIter + Slice<RangeFrom<usize>>,
{
    move |mut inp: I| {
        while inp.input_len() > 0 {
            if let Ok(t) = g.parse(inp.clone()) {
                return Ok(t);
            } else {
                inp = inp.slice(1..);
            }
        }
        fail(inp)
    }
}

/// Greedily collect matches for the first parser, discarding anything that does not match until the second parser succeeds
/// Returns Err if the second parser does not succeed before EOF
fn find_many_till<I, O, P>(
    mut f: impl FnMut(I) -> IResult<I, O>,
    mut g: impl FnMut(I) -> IResult<I, P>,
    inp: I,
) -> IResult<I, (Vec<O>, P)>
where
    I: Clone + InputTake + InputLength + InputIter,
{
    let mut all_found = Vec::new();
    let mut inp = inp;
    loop {
        match g(inp.clone()) {
            Err(_) => {
                if let Ok((rest, found)) = f(inp.clone()) {
                    all_found.push(found);
                    inp = rest;
                } else {
                    let (rest, _) = take(1usize)(inp)?;
                    inp = rest;
                }
            }
            Ok((rest, stop)) => return Ok((rest, (all_found, stop))),
        }
    }
}

/// finds all non-overlapping matches for a given parser until the input is exhausted
fn _find_all<I, O, E, F>(mut pat: F) -> impl FnMut(I) -> IResult<I, Vec<O>, E>
where
    I: InputIter + InputTake + Clone + InputLength + Slice<RangeFrom<usize>>,
    F: Parser<I, O, E>,
    E: ParseError<I>,
{
    move |mut i: I| {
        let mut all_found = Vec::new();
        loop {
            match pat.parse(i.clone()) {
                Ok((rest, found)) => {
                    i = rest;
                    all_found.push(found);
                }
                Err(_) => {
                    if i.input_len() > 0 {
                        i = i.slice(1..);
                    } else {
                        return Ok((i, all_found));
                    }
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {

    use super::*;
    use nom::combinator::not;
    use nom::multi::many0_count;
    use nom::Offset;

    #[test]
    fn test_find_all() {
        let src = "needle fskjnfonwofinwfeneedlekfsdond;nsdf;ondsfneedlefsmfok";
        let (rest, found) = _find_all(tag::<&str, &str, ()>("needle"))(src).unwrap();
        assert_eq!(found, vec!["needle", "needle", "needle"]);
        assert_eq!(rest, "");
    }

    #[test]
    fn test_find_many_till() {
        let src = "abcdefsba bababaaaaabbbbbabababab";
        let (rest, (found, end)) =
            find_many_till(alt((tag("a"), tag("b"))), tag("baba"), src).unwrap();
        assert_eq!(found, vec!["a", "b", "b", "a"]);
        assert_eq!(end, "baba");
        assert_eq!(rest, "baaaaabbbbbabababab");
        assert!(find_many_till(tag("a"), tag("b"), "aaaaaaaacccccc").is_err());
    }

    #[test]
    fn test_find_next() {
        let src = "foovrfs barand more";
        let needle = tag::<_, _, ()>("bar");
        let mut f_next = many0_count(not(&needle).and(take(1usize))).and(&needle);
        let (rest, res) = find_next(tag("bar"))(src).unwrap();
        let (rest2, res2) = f_next.parse(src).unwrap();
        assert_eq!(res, "bar");
        assert_eq!(rest, "and more");
        assert_eq!(res2, (8, "bar"));
        assert_eq!(rest2, "and more");
        assert!(find_next(tag("foo"))("haystack with no match").is_err());
    }

    #[test]
    fn test_parse_link() {
        let undertest = "[Foo](bar) and more";
        let expected = TokenType::Link(Link {
            text: "Foo",
            path: "bar",
        });
        let (rest, _parsed) = assert_parse_token_with_inner(undertest, parse_link, expected);
        assert_eq!(*rest.fragment(), " and more");
    }

    #[test]
    fn test_parse_to_directive() {
        let undertest =
            "To really `bake` a `cake` with `--flour`, `ignoreme` `--butter` and `--milk`: \r\nand more";
        let expected = TokenType::ToDirective(ToDirective {
            callpath: vec!["bake", "cake"],
            args: vec!["flour", "butter", "milk"],
        });
        let (rest, _) = assert_parse_token_with_inner(undertest, parse_to_directive, expected);
        assert_eq!(*rest.fragment(), "and more");
        assert!(parse_to_directive("To do nothing is no fun...\n".into()).is_err());
        assert!(parse_to_directive(
            "To have a `valid` to-directive without trailing newline is meaningless".into()
        )
        .is_err());
        let undertest = "To `have` `cake` without args is okay\n";
        let expected = TokenType::ToDirective(ToDirective {
            callpath: vec!["have", "cake"],
            args: vec![],
        });
        assert_parse_token_with_inner(undertest, parse_to_directive, expected);
    }

    #[test]
    fn test_parse_code_block() {
        let undertest = r#"```sh foo_bar (object) annot1 annot2
echo "I am a {{ banana }}"
```
and more"#;
        let expected = TokenType::CodeBlock(CodeBlock {
            executor: Some("sh"),
            name: Some("foo_bar"),
            type_hint: Some("object"),
            annotations: vec!["annot1", "annot2"],
            code: "echo \"I am a {{ banana }}\"\n",
        });
        let (rest, _) = assert_parse_token_with_inner(undertest, parse_code_block, expected);
        assert_eq!(*rest.fragment(), "\nand more");
    }

    #[test]
    fn test_parse_ident() {
        let undertest = "`_foo_bar.?bafroo\\` and this is whats left";
        let expected = TokenType::Ident("_foo_bar.?bafroo\\");
        let (rest, _) = assert_parse_token_with_inner(&undertest, parse_single_ident, expected);
        assert_eq!(*rest.fragment(), " and this is whats left");
    }

    #[test]
    fn test_parse_heading() {
        let undertest = "#### \t How To Bake a Cake\t  \r\nand live to tell the tale";
        let expected = TokenType::Heading(Heading {
            level: 4,
            text: "How To Bake a Cake",
        });
        let (rest, parsed) = assert_parse_token_with_inner(&undertest, parse_heading, expected);
        assert_eq!(
            parsed.position.offset(&rest),
            undertest.len() - rest.fragment().len()
        );
        assert_eq!(*rest.fragment(), "and live to tell the tale");
    }

    fn assert_parse_token_with_inner<'a>(
        src: &'a str,
        parser: impl Fn(Span) -> IResult<Span, Token>,
        expected_inner: TokenType,
    ) -> (Span<'a>, Token<'a>) {
        let undertest = LocatedSpan::new(src);
        let (rest, parsed) = parser(undertest).unwrap();
        let _f: Result<
            (LocatedSpan<&str>, Token<'_>),
            nom::Err<nom::error::Error<LocatedSpan<&str>>>,
        > = parser(undertest);
        assert_eq!(parsed.inner, expected_inner);
        (rest, parsed)
    }
}
