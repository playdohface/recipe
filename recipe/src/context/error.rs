use super::Executor;
use crate::parser::SelectionPath;
use thiserror::Error;

#[derive(Debug)]
pub enum ExecutionError {
    Generic,
    TemplateError,
    ExecutorFailed((Executor, std::io::Error)),
    MissingExecutor,
    ValueNotFound,
    InvalidPath,
}
