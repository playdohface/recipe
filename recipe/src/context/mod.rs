use std::{
    collections::{BTreeMap, HashMap},
    process::Output,
};

mod error;

use error::ExecutionError;
use handlebars::Handlebars;
use serde_json::{json, Value};

use crate::parser::CodeBlock;

trait Template {
    fn template(&self) -> &str;
}

impl<'a> Template for CodeBlock<'a> {
    fn template(&self) -> &str {
        self.code
    }
}

/// AppContext serves as the anchor for an invocation of recipe
#[derive(Debug, Default)]
pub struct AppContext<'a> {
    commands: Commands<'a>,
    ctx: Context,
}
impl<'a> AppContext<'a> {
    pub fn new() -> Self {
        AppContext {
            commands: Commands {
                inner: BTreeMap::new(),
            },
            ctx: Context {
                root: HashMap::new(),
                executors: Executors::default(),
            },
        }
    }
    pub fn register_command(&mut self, name: &'a str, command: Command<'a>) {
        self.commands.inner.insert(name.to_owned(), command);
    }
    pub fn register_variable(&mut self, name: &str, value: Value) {
        self.ctx.root.insert(name.to_owned(), value);
    }
    pub fn register_executor(&mut self, name: &str, executor: Executor) {
        self.ctx.executors.inner.insert(name.to_owned(), executor);
    }
}

/// Wraps all the loaded Commands
#[derive(Debug, Default)]
pub struct Commands<'a> {
    inner: BTreeMap<String, Command<'a>>,
}

#[derive(Debug)]
pub struct Executors {
    inner: HashMap<String, Executor>,
}
impl Default for Executors {
    fn default() -> Self {
        let mut inner = HashMap::new();
        inner.insert("sh".to_owned(), Executor::shell());
        Executors { inner }
    }
}
impl Executors {
    fn run(&self, name: &str, code: String) -> Result<Output, ExecutionError> {
        if let Some(exec) = self.inner.get(name) {
            exec.run(code)
        } else {
            Executor::generic(name).run(code)
        }
    }
}

/// Something that can execute a script of a certain type
/// May need tweaking to accomodate for different types of scripts
#[derive(Debug, Clone)]
pub struct Executor {
    program: String,
    args: Vec<String>,
}
impl Executor {
    fn run(&self, code: String) -> Result<Output, ExecutionError> {
        std::process::Command::new(&self.program)
            .args(&self.args)
            .arg(&code)
            .output()
            .map_err(|e| ExecutionError::ExecutorFailed((self.clone(), e)))
    }
    pub fn shell() -> Self {
        Executor {
            program: "/bin/sh".to_owned(),
            args: vec!["-c".to_owned()],
        }
    }
    fn generic(executable: &str) -> Self {
        Executor {
            program: executable.to_owned(),
            args: vec!["-".to_owned()],
        }
    }
}

/// Wraps the context passed to each command-invocation
#[derive(Debug, Default)]
pub struct Context {
    /// the root of the template context
    root: HashMap<String, serde_json::Value>,
    /// the loaded Executors
    executors: Executors,
}
impl Context {
    fn register_execution(&mut self, output: Output, type_hint: Option<&str>) {
        //TODO: implement type converters
        let stdout =
            json!(String::from_utf8(output.stdout).unwrap_or("TODO: Not UTF8!".to_owned()));
        let stderr =
            json!(String::from_utf8(output.stderr).unwrap_or("TODO: Not UTF8!".to_owned()));
        let statuscode = json!(output.status.code());
        self.root.insert("output".to_owned(), stdout);
        self.root.insert("error".to_owned(), stderr);
        self.root.insert("status".to_owned(), statuscode);
    }
}

#[derive(Debug)]
/// Represents a single loaded command
pub enum Command<'a> {
    Composite(Vec<Command<'a>>),
    CodeBlock(CodeBlock<'a>),
}
impl<'a> Command<'a> {
    fn execute(&self, ctx: &mut Context) -> Result<(), error::ExecutionError> {
        match self {
            Self::Composite(commands) => {
                for cmd in commands {
                    cmd.execute(ctx)?
                }
                Ok(())
            }
            Self::CodeBlock(codeblock) => {
                let executor = codeblock.executor.ok_or(ExecutionError::MissingExecutor)?;
                let template = codeblock.code;
                //TODO we may not want to construct this every time...
                let handlebars = Handlebars::new();
                let code = handlebars
                    .render_template(template, &ctx.root)
                    .map_err(|e| ExecutionError::TemplateError)?;
                let output = ctx.executors.run(executor, code)?;
                ctx.register_execution(output, codeblock.type_hint);
                Ok(())
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn shell_code_block(code: &str) -> CodeBlock {
        CodeBlock {
            executor: Some("sh".into()),
            name: None,
            type_hint: None,
            annotations: vec![],
            code: code.into(),
        }
    }

    #[test]
    fn can_execute_and_interpolate_codeblock() {
        let codeblock = shell_code_block("echo \"Hello, World!\"\n");
        let command = Command::CodeBlock(codeblock);
        let mut context = Context {
            root: HashMap::new(),
            executors: Executors::default(),
        };
        let res = command.execute(&mut context);
        assert!(res.is_ok());
        assert_eq!(
            context.root.get("output").unwrap(),
            &json!("Hello, World!\n")
        );
        let codeblock = shell_code_block("echo \"Output was: {{ output }}\"\n");
        let command = Command::CodeBlock(codeblock);
        let res = command.execute(&mut context);
        assert!(res.is_ok());
        assert_eq!(
            context.root.get("output").unwrap(),
            &json!("Output was: Hello, World!\n\n")
        );
    }

    #[test]
    fn can_run_generic_script() {
        let command = Command::CodeBlock(CodeBlock {
            executor: Some("node".into()),
            name: None,
            type_hint: None,
            annotations: vec![],
            code: "console.log(\"Hello from node\");".into(),
        });
        let mut context = Context {
            root: HashMap::new(),
            executors: Executors::default(),
        };
        let res = command.execute(&mut context);
        dbg!(context);
    }
}
