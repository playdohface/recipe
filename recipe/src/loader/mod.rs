use std::path::Path;

use convert_case::{Case, Casing};
use itertools::Itertools;

use crate::context::{Command, Commands};
use crate::parser::{Tokens, TokenType};
use crate::parser::parse_to_directive_inner;

/// Load the Context
pub fn load(root_path: &Path) -> anyhow::Result<Commands> {
    let recipe = load_file(root_path)?;
    let mut tokens = Tokens::new();
    tokens.load(recipe, root_path, None);
    let mut commands = Commands::new();
    while let Some(token) = tokens.next() {
        use crate::parser::Keyword::*;
        use TokenType::*;
        match token.inner() {
            Link(link) => load_link(&mut tokens, link.path),
            Block(code) => load_block(&mut tokens, &mut commands, code),
            Keyword(To) => load_to_directive(&mut tokens, &mut commands),
            Keyword(_keyword) => todo!(),
            Inline(_) | Heading(_) | Newline => continue,
        }
    }
    Ok(commands)
}

fn load_link<'a>(tokens: &mut Tokens<'a>, origin_path: &'a str) {
    if origin_path.is_empty() {
        return;
    }
    let (mut file_path, heading_slug) = match origin_path.split('#').collect::<Vec<_>>().as_slice()
    {
        [file_path] => (*file_path, None),
        [file_path, heading_slug] => (*file_path, Some(*heading_slug)),
        _ => return,
    };
    if file_path.is_empty() {
        file_path = tokens.current_path().and_then(|p|p.to_str()).unwrap_or_default();
    }
    if let Ok(contents) = load_file(file_path) {
        tokens.load(contents, file_path.as_ref(), heading_slug);
    }
}

/// Resolve a path and return the contents of the file
/// TODO - There will be relative paths and default locations
pub fn load_file(path: impl AsRef<Path>) -> Result<String, std::io::Error> {
    std::fs::read_to_string(path)
}

fn load_block(_tokens: &mut Tokens, _commands: &mut Commands, _code: &str) {
    todo!()
}

fn load_to_directive<'a>(tokens: &mut Tokens<'a>, commands: &mut Commands<'a>) {
    let mut names = Vec::new();
    for token in tokens.by_ref() {
        match token.inner() {
            TokenType::Inline(inline) => {
                names.push(*inline);
            }
            TokenType::Newline => break,
            _ => {
                //TODO Output a warning. Do we break or continue here?
                continue;
            }
        }
    }
    let name = names.iter().map(|n| n.to_case(Case::Kebab)).join("-");
    if name.is_empty() {

        //return; //TODO errors/warnings
    }
    let parsed_commands = parse_to_directive_inner(tokens);
    commands.register_command(name, Command::Composite(parsed_commands));
}

#[cfg(test)]
mod tests {}
