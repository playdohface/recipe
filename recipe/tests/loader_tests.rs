#[cfg(test)]
mod loader_tests {
    use std::fs::File;
    use std::io::Write;
    use itertools::Itertools;
    use serde_json::json;
    use tempfile::{TempDir, tempdir};
    use recipe::context::{AppContext, Command};
    use recipe::loader::load;
    use recipe::parser::{CodeBlock, Heading, Keyword, Selection, SelectionPath, SetDirective, TokenType, Tokens, Val};

    #[test]
    fn load_simple_recipe() {
        let contents = r#"
# Heading 1
To `bake` a `cake`
Set the `topping` to `"cherry"`
All you need to do is run the following command:
```sh
echo {{ topping }}
```
"#;
        let (base_path, _dir, _file) = write_temp_file("Recipe.md", contents);
        let loaded = load(&base_path).unwrap();

        let set_cherry = Command::SetDirective(SetDirective { variable: SelectionPath (vec![Selection::Key("topping")]), value: Val::Literal(json!("cherry")) });
        let echo_topping = Command::CodeBlock(CodeBlock { executor: Some("sh"), name: None, type_hint: None, annotations: vec![], code: "echo {{ topping }}\n" });
        let bake_cake = Command::Composite(vec![set_cherry, echo_topping]);
        let mut expected = AppContext::new();
        expected.register_command("bake cake".to_string(), bake_cake);
        assert_eq!(loaded, expected);

    }

    #[test]
    fn load_tokens() {
        let contents = r#"# Heading 1
To `bake` a `cake`
Set `foo` to `"bar"`
```sh
echo {{ topping }}
```
"#;
        let mut tokens = Tokens::new();
        tokens.load(contents.to_string(), "foo", None);
        assert_eq!(tokens.map(|t|t.inner().clone()).collect_vec(), vec![
            TokenType::Heading(Heading{ level: 1, text: "Heading 1" }),
            TokenType::Keyword(Keyword::To),
            TokenType::Inline("bake"),
            TokenType::Inline("cake"),
            TokenType::Newline,
            TokenType::Keyword(Keyword::Set),
            TokenType::Inline("foo"),
            TokenType::Inline("\"bar\""),
            TokenType::Newline,
            TokenType::Block("sh\necho {{ topping }}\n"),
            TokenType::Newline
        ]);
    }

    fn write_temp_file(filename: &str, contents: &str) -> (String, TempDir, File) {
        let dir = tempdir().unwrap();
        let file_path = dir.path().join(filename);
        let mut file = File::create(&file_path).unwrap();
        write!(file, "{}", contents).unwrap();
        (file_path.as_os_str().to_str().unwrap().to_owned(), dir, file)
    }
}