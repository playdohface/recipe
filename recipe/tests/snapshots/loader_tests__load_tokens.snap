---
source: tests/loader_tests.rs
expression: "tokens.collect::<Vec<_>>()"
snapshot_kind: text
---
- start: 0
  end: 12
  inner:
    Heading:
      level: 1
      text: Heading 1
- start: 12
  end: 15
  inner:
    Keyword: To
- start: 16
  end: 21
  inner:
    Inline: bake
- start: 25
  end: 30
  inner:
    Inline: cake
- start: 30
  end: 31
  inner: Newline
- start: 31
  end: 35
  inner:
    Keyword: Set
- start: 36
  end: 40
  inner:
    Inline: foo
- start: 45
  end: 51
  inner:
    Inline: "\"bar\""
- start: 51
  end: 52
  inner: Newline
- start: 55
  end: 80
  inner:
    Block: "sh\necho {{ topping }}\n"
- start: 80
  end: 81
  inner: Newline
